README
SDM4 importer til sikrede

Changelog.md

## sikredeimporter 4.3
* Fixed NSPSUPPORT-154 Grundet ændringer i input formatet kan SIkraftDatoYderGl indeholde udelukkende nuller i stedet
  for en korrekt dato, dette omgås nu ved at indsætte epoch dato i stedet for at fejle.

## sikredeimporter 4.4
*  Løser NSPSUPPORT-167 fejlende import.
   Det er løst ved at tillade flere samtidige behandlingsrelationer mellem patienter og læger.

## sikredeimporter 4.5
*  SDM-19 Sikrede register mangler index på CPR-nummer - løst med en simpel database migrering.

## sikredeimporter 4.6
*  Opdateret SDM4 depencencies
*  SDM-5 SLA-log fra SDM4-importere følger ikke standarden
*  Tilføjet kopi register view, så kopi register service maps nu bliver oprettet automatisk
*  Tilføjet nyt index der er gør krs kopiering hurtigere.


src/test/java/dk/nsi/sdm4/sikrede/brs/BrsUpdaterTest.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sikrede.brs;

/**
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * Contributor(s): Contributors are attributed in the source code
 * where applicable.
 *
 * The Original Code is "Stamdata".
 *
 * The Initial Developer of the Original Code is Trifork Public A/S.
 *
 * Portions created for the Original Code are Copyright 2011,
 * Lægemiddelstyrelsen. All Rights Reserved.
 *
 * Portions created for the FMKi Project are Copyright 2011,
 * National Board of e-Health (NSI). All Rights Reserved.
 */

import dk.nsi.sdm4.core.persistence.recordpersister.Record;
import dk.nsi.sdm4.sikrede.config.SikredeimporterApplicationConfig;
import dk.nsi.sdm4.testutils.TestDbConfiguration;
import org.joda.time.DateTime;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.support.rowset.SqlRowSet;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Date;
import java.sql.SQLException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(SpringJUnit4ClassRunner.class)
@Transactional
@ContextConfiguration(classes = {SikredeimporterApplicationConfig.class, TestDbConfiguration.class})
public class BrsUpdaterTest {
    private static final DateTime ASSIGNED_FROM = new DateTime(2011, 10, 10, 0, 0);
    private static final DateTime ASSIGNED_TO = new DateTime(2011, 11, 10, 0, 0);

    private String examplePatientCpr;
    private String exampleDoctorOrganisationIdentifier;

    @Autowired
    private JdbcTemplate jdbcTemplate;
    @Autowired
    private BrsUpdater brsUpdater;

    @Before
    public void setup() {
        examplePatientCpr = "1234567890123456789012345678901234567890";
        exampleDoctorOrganisationIdentifier = "123456";

    }

    @Test
    public void testHashCpr() {
        String hashedCpr = BrsUpdater.hashCpr("1234567890");
        assertEquals(40, hashedCpr.length());
    }

    @Test
    public void testParseSikredeRecordDate() {
	    Record dateRecord = new Record();
	    dateRecord.put("date", "20110507");

	    DateTime parsedDate = BrsUpdater.parseSikredeRecordDate(dateRecord, "date");

	    DateTime expectedDate = new DateTime(2011, 5, 7, 0, 0, 0);
        assertEquals(expectedDate, parsedDate);
    }

    @Test
    public void parseSikredeRecordDateEpochIfEmpty() {
        Record dateRecord = new Record();
        dateRecord.put("date", "00000000");

        DateTime parsedDate = BrsUpdater.parseSikredeRecordDateDefaultDateIfEmpty(dateRecord, "date");
        assertEquals(1970, parsedDate.getYear());

        dateRecord = new Record();
        dateRecord.put("date", "20110507");
        parsedDate = BrsUpdater.parseSikredeRecordDateDefaultDateIfEmpty(dateRecord, "date");
        DateTime expectedDate = new DateTime(2011, 5, 7, 0, 0, 0);
        assertEquals(expectedDate, parsedDate);
    }
/*
    @Test
    public void testOpenRelationshipExistsWhenNoRecords() throws SQLException {
        assertEquals(BrsUpdater.NO_EXISTING_RELATIONSHIP, brsUpdater.openRelationshipExists(examplePatientCpr,
                exampleDoctorOrganisationIdentifier));
    }

    @Test
    public void testOpenRelationshipExistsWhenRecordWhichHasOtherDoctorOrganisationIdExists() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, "654321", ASSIGNED_FROM, null);
        assertEquals(BrsUpdater.NO_EXISTING_RELATIONSHIP, brsUpdater.openRelationshipExists(examplePatientCpr,
                exampleDoctorOrganisationIdentifier));
    }

    @Test
    public void testOpenRelationshipExistsWhenRecordThatIsClosedIsFoundInDatabase() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM,
                ASSIGNED_TO);
        assertEquals(BrsUpdater.NO_EXISTING_RELATIONSHIP, brsUpdater.openRelationshipExists(examplePatientCpr,
                exampleDoctorOrganisationIdentifier));
    }

    @Test
    public void testOpenRelationshipExistsWhenMatchingRecordInDatabase() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, null);
        assertTrue(BrsUpdater.NO_EXISTING_RELATIONSHIP != brsUpdater.openRelationshipExists(examplePatientCpr,
                exampleDoctorOrganisationIdentifier));
    }

    @Test
    public void testCloseRelationship() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, null);
        long primaryKeyOfOpenRelationship = brsUpdater.openRelationshipExists(examplePatientCpr,
                exampleDoctorOrganisationIdentifier);
        brsUpdater.closeRelationship(primaryKeyOfOpenRelationship, ASSIGNED_TO);
        assertClosedRelationship(primaryKeyOfOpenRelationship);
    }

    @Test
    public void testUpdateExistingRelationshipWhenDoctorDoesNotExist() throws SQLException {
        brsUpdater.updateExistingRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM,
                ASSIGNED_TO);
        assertClosedRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier);
    }

    @Test
    public void testUpdateExistingRelationshipWhenDoctorExistWithOpenRelationship() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, null);
        assertRecordExists(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM);
        brsUpdater.updateExistingRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM
                .minusDays(1), ASSIGNED_TO);
        assertRecordExists(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, ASSIGNED_TO);
    }

    @Test
    public void testUpdateExistingRelationshipWhenDoctorExistWithClosedRelationship() throws SQLException {
        brsUpdater.insertRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM,
                ASSIGNED_TO);
        assertRecordExists(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, ASSIGNED_TO);
        brsUpdater.updateExistingRelationship(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_TO
                .plusYears(2), ASSIGNED_TO.plusYears(3));
        assertRecordExists(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_FROM, ASSIGNED_TO);
        assertRecordExists(examplePatientCpr, exampleDoctorOrganisationIdentifier, ASSIGNED_TO.plusYears(2),
                ASSIGNED_TO.plusYears(3));
    }

    // //////////////////////

    private void assertRecordExists(String patientCpr, String doctorOrganisationIdentifier, DateTime assignedFrom, DateTime assignedTo) {

        String sql = "SELECT * FROM AssignedDoctor WHERE patientCpr = ? AND doctorOrganisationIdentifier = ? AND assignedFrom = ? AND assignedTo = ?";
        SqlRowSet rs = jdbcTemplate.queryForRowSet(sql, patientCpr, doctorOrganisationIdentifier, new Date(assignedFrom.getMillis()), new Date(assignedTo.getMillis()));
        assertTrue(rs.next());
        
    }

    private void assertRecordExists(String patientCpr, String doctorOrganisationIdentifier, DateTime assignedFrom) {
        
        String sql = "SELECT * FROM AssignedDoctor WHERE patientCpr = ? AND doctorOrganisationIdentifier = ? AND assignedFrom = ? AND assignedTo IS NULL";
        SqlRowSet rs = jdbcTemplate.queryForRowSet(sql, patientCpr, doctorOrganisationIdentifier, new Date(assignedFrom.getMillis()));
        assertTrue(rs.next());
    }

    private void assertClosedRelationship(long pk) {
        
        String sql = "SELECT * FROM AssignedDoctor WHERE pk = ? AND assignedTo IS NOT NULL";
        SqlRowSet rs = jdbcTemplate.queryForRowSet(sql, pk);
        assertTrue(rs.next());
    }

    private void assertClosedRelationship(String patientCpr, String doctorOrganisationIdentifier) {

        String sql = "SELECT * FROM AssignedDoctor WHERE patientCpr = ? AND doctorOrganisationIdentifier = ? AND assignedTo IS NOT NULL";
        SqlRowSet rs = jdbcTemplate.queryForRowSet(sql, patientCpr, doctorOrganisationIdentifier);
        assertTrue(rs.next());
    }                                 */
}


src/test/java/dk/nsi/sdm4/sikrede/parser/RecordGenerator.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sikrede.parser;

import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.RecordFieldType.ALPHANUMERICAL;
import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.RecordFieldType.DECIMAL10_3;
import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.RecordFieldType.NUMERICAL;
import static org.junit.Assert.assertTrue;

import dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification;
import dk.nsi.sdm4.core.persistence.recordpersister.Record;
import dk.nsi.sdm4.core.persistence.recordpersister.RecordSpecification;

public class RecordGenerator {
    private RecordSpecification recordSpecification;

    public RecordGenerator(RecordSpecification recordSpecification) {
        this.recordSpecification = recordSpecification;
    }

    public String stringFromRecords(Record record) {
        if (!recordSpecification.conformsToSpecifications(record)) {
            throw new IllegalArgumentException("Sikrede record does not conform to specification");
        }

        StringBuilder builder = new StringBuilder();
        for (FieldSpecification fieldSpecification : recordSpecification.getFieldSpecs()) {
	        try {
		        if (fieldSpecification.type == ALPHANUMERICAL) {
			        String value = (String) record.get(fieldSpecification.name);
			        builder.append(prefixPadding(' ', fieldSpecification.length - value.length()));
			        builder.append(value);
		        } else if (fieldSpecification.type == NUMERICAL) {
			        String value = Long.toString((Long) record.get(fieldSpecification.name));
			        builder.append(prefixPadding('0', fieldSpecification.length - value.length()));
			        builder.append(value);
		        } else {
			        throw new AssertionError("Missing implementation for type " + fieldSpecification.type);
		        }
	        } catch (RuntimeException e) {
		        throw new RuntimeException("While getting " + fieldSpecification.name + " from record " + record, e);
	        }
        }
        return builder.toString();
    }

    public String stringFromIncompleteRecord(Record record) {
        for (FieldSpecification fieldSpecification : recordSpecification.getFieldSpecs()) {
            if (!record.containsKey(fieldSpecification.name)) {
                if (fieldSpecification.type == ALPHANUMERICAL) {
                    record = record.put(fieldSpecification.name, "");
                } else if (fieldSpecification.type == NUMERICAL) {
                    record = record.put(fieldSpecification.name, 0L);
                } else if (fieldSpecification.type == DECIMAL10_3) {
	                record = record.put(fieldSpecification.name, 0.0);
                } else {
                    throw new AssertionError("Missing implementation for type " + fieldSpecification.type);
                }
            }
        }

        return stringFromRecords(record);
    }

    public String stringRecordFromIncompleteSetOfFields(Object... keysAndValues) {
        return stringFromIncompleteRecord(createRecord(keysAndValues));
    }

    public static Record createRecord(Object... keysAndValues) {
        Record record = new Record();

        assertTrue(keysAndValues.length % 2 == 0);

        for (int i = 0; i < keysAndValues.length; i += 2) {
            String key = (String) keysAndValues[i];
            Object value = keysAndValues[i + 1];
            record = record.put(key, value);
        }

        return record;
    }

    private String prefixPadding(char paddingElement, int n) {
        String padding = "";
        for (int i = 0; i < n; i++) {
            padding += paddingElement;
        }
        return padding;
    }
}


src/test/java/dk/nsi/sdm4/sikrede/parser/SikredeFieldsTest.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sikrede.parser;

import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.RecordFieldType.ALPHANUMERICAL;
import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.RecordFieldType.NUMERICAL;
import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.field;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification;
import org.junit.Test;

import com.google.common.collect.Iterables;

import dk.nsi.sdm4.core.persistence.recordpersister.Record;
import dk.nsi.sdm4.core.persistence.recordpersister.RecordSpecification;
import dk.nsi.sdm4.sikrede.recordspecs.SikredeRecordSpecs;

public class SikredeFieldsTest {
    @Test
    public void testCorrectNumberOfFields() {
        Iterable<FieldSpecification> fieldSpecs = SikredeRecordSpecs.ENTRY_RECORD_SPEC.getFieldSpecs();
        assertEquals(Iterables.size(fieldSpecs), 48);
    }

    @Test
    public void testCorrectNumberOfAlphanumericalFields() {
        Iterable<FieldSpecification> fieldSpecs = SikredeRecordSpecs.ENTRY_RECORD_SPEC.getFieldSpecs();

        int alphanumericalFields = 0;

        for (FieldSpecification spec : fieldSpecs) {
            if (spec.type == ALPHANUMERICAL) {
                alphanumericalFields++;
            }
        }

        assertEquals(47, alphanumericalFields);
    }

    @Test
    public void testCorrectNumberOfNumericalFields() {
        Iterable<FieldSpecification> fieldSpecs = SikredeRecordSpecs.ENTRY_RECORD_SPEC.getFieldSpecs();

        int numericalFields = 0;
        for (FieldSpecification spec : fieldSpecs) {
            if (spec.type == NUMERICAL) {
                numericalFields++;
            }
        }

        assertEquals(1, numericalFields);
    }

    @Test
    public void testCorrectAcceptedTotalLineLength() {
        RecordSpecification exampleRecordSpecification = RecordSpecification.createSpecification("SikredeGenerated",
                "Foo", field("Foo", 10, false).numerical(), field("Bar", 32, false));
        assertEquals(42, exampleRecordSpecification.acceptedTotalLineLength());
    }

    @Test
    public void testCorrectAcceptedTotalLineLengthForSingleton() {
        assertEquals(634, SikredeRecordSpecs.ENTRY_RECORD_SPEC.acceptedTotalLineLength());
    }

    @Test
    public void testConformsToSchemaSpecification() {
        RecordSpecification exampleRecordSpecification = RecordSpecification.createSpecification("SikredeGenerated",
                "Foo", field("Foo", 10, false).numerical(), field("Bar", 32, false));

        Record correctValues = RecordGenerator.createRecord("Foo", 42, "Bar", "12345678901234567890123456789012");
        Record correctValuesWhereBarIsShorter = RecordGenerator.createRecord("Foo", 42, "Bar",
                "123456789012345678901234567890");
        Record missingFoo = RecordGenerator.createRecord("Bar", "12345678901234567890123456789012");
        Record fooIsNotNumerical = RecordGenerator
                .createRecord("Foo", "Baz", "Bar", "12345678901234567890123456789012");
        Record barIsTooLong = RecordGenerator
                .createRecord("Foo", 42, "Bar", "1234567890123456789012345678901234567890");

        assertTrue(exampleRecordSpecification.conformsToSpecifications(correctValues));
        assertTrue(exampleRecordSpecification.conformsToSpecifications(correctValuesWhereBarIsShorter));
        assertFalse(exampleRecordSpecification.conformsToSpecifications(missingFoo));
        assertFalse(exampleRecordSpecification.conformsToSpecifications(fooIsNotNumerical));
        assertFalse(exampleRecordSpecification.conformsToSpecifications(barIsTooLong));
    }
}


src/main/java/dk/nsi/sdm4/sikrede/brs/BrsUpdater.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sikrede.brs;

import org.apache.log4j.Logger;
import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;

import com.google.common.base.Preconditions;

import dk.nsi.sdm4.core.persistence.recordpersister.Record;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Date;
import java.sql.SQLException;
import java.util.List;

public class BrsUpdater {
    
    @Autowired
    private JdbcTemplate jdbcTemplate;

    private static final Logger log = Logger.getLogger(BrsUpdater.class);

    public BrsUpdater() {
    }

    public void updateRecord(Record record) throws SQLException {
        String hashedCpr = hashCpr((String) record.get("CPRnr"));

        updateExistingRelationship(hashedCpr,
                (String) record.get("SYdernrGl"),
                // The input was changed so we can now get a SIkraftDatoYderGl containing all zeroes, this indicates
                // that the person always had the same doctor, so we return epoch as nulls is not allowed in db.
                parseSikredeRecordDateDefaultDateIfEmpty(record, "SIkraftDatoYderGl"),
                parseSikredeRecordDate(record, "SIkraftDatoYder"));
        insertRelationship(hashedCpr,
                (String) record.get("SYdernr"),
                parseSikredeRecordDate(record, "SIkraftDatoYder"),
                null);
    }

    void updateExistingRelationship(String patientCpr,
                                    String doctorOrganisationIdentifier,
                                    DateTime assignedFrom,
                                    DateTime assignedTo) throws SQLException {
        List<Long> existingRelations = openRelationshipsExists(patientCpr, doctorOrganisationIdentifier);
        if (existingRelations.size() > 0) {
            log.debug("Found " + existingRelations.size() + " existing relationsships");
        }
        if (existingRelations == null || existingRelations.size() == 0) {
            insertRelationship(patientCpr, doctorOrganisationIdentifier, assignedFrom, assignedTo);
        } else {
            closeRelationships(existingRelations, assignedTo);
        }
    }

    List<Long> openRelationshipsExists(String patientCpr, String doctorOrganisationIdentifier) throws SQLException {
        String querySql = "SELECT pk FROM AssignedDoctor WHERE patientCpr = ? AND doctorOrganisationIdentifier = ? AND assignedTo IS NULL";
        try {
            return jdbcTemplate.queryForList(querySql, Long.class, patientCpr, doctorOrganisationIdentifier);
        } catch (EmptyResultDataAccessException norelation) {
            return null;
        }
    }

    void closeRelationships(List<Long> keys, DateTime assignedTo) throws SQLException {
        String updateSql = "UPDATE AssignedDoctor SET assignedTo = ? WHERE pk = ?";
        for (Long key : keys) {
            jdbcTemplate.update(updateSql, new Date(assignedTo.getMillis()), key);
        }
    }

    void insertRelationship(String patientCpr, String doctorOrganisationIdentifier, DateTime assignedFrom,
            DateTime assignedTo) throws SQLException {
        String insertSql = "INSERT INTO AssignedDoctor (patientCpr, doctorOrganisationIdentifier, assignedFrom, assignedTo, reference) VALUES (?, ?, ?, ?, ?)";

        Date assignedToDate = null;
        if (assignedTo != null) {
            assignedToDate = new Date(assignedTo.getMillis());
        }

        jdbcTemplate.update(insertSql, patientCpr, doctorOrganisationIdentifier, new Date(assignedFrom.getMillis()),
                assignedToDate, new DateTime().toString());
    }

    static String hashCpr(String cpr) {
        Preconditions.checkArgument(cpr.length() == 10);
        try {
            return hash(cpr);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Parse a date, in case the input is all zeroes, epoch is returned.
     * @param record
     * @param key name of the key in the records that contains the date
     * @return
     */
    static DateTime parseSikredeRecordDateDefaultDateIfEmpty(Record record, String key) {
        String date = (String) record.get(key);
        if (date.equals("00000000")) {
            return new DateTime(0l);
        } else {
            return parseSikredeRecordDate(record, key);
        }
    }

    static DateTime parseSikredeRecordDate(Record record, String key) {
	    String date = (String) record.get(key);
	    final int expectedLength = 8;
	    Preconditions.checkArgument(date.length() == expectedLength,
                "Date '%s' with key %s from record %s has wrong length, expected %s", date, key, record, expectedLength);
        return new DateTime(
		        Integer.parseInt(date.substring(0, 4)),
		        Integer.parseInt(date.substring(4, 6)),
		        Integer.parseInt(date.substring(6, 8)),
		        0, 0, 0);
    }

    private static final String SHA_1 = "SHA-1";

    private static String hash(String string) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance(SHA_1);
        digest.reset();
        byte[] bytes = digest.digest(string.getBytes());
        return getHex(bytes);
    }

    private static final String HEXES = "0123456789ABCDEF";

    private static String getHex(byte[] raw) {
        if (raw == null) {
            return null;
        }
        final StringBuilder hex = new StringBuilder(2 * raw.length);
        for (final byte b : raw) {
            hex.append(HEXES.charAt((b & 0xF0) >> 4)).append(HEXES.charAt((b & 0x0F)));
        }
        return hex.toString();
    }

}


src/main/java/dk/nsi/sdm4/sikrede/recordspecs/SikredeRecordSpecs.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sikrede.recordspecs;

import dk.nsi.sdm4.core.persistence.recordpersister.RecordSpecification;

import static dk.nsi.sdm4.core.persistence.recordpersister.FieldSpecification.field;

public final class SikredeRecordSpecs
{
    protected SikredeRecordSpecs() {}
    
    public static final RecordSpecification START_RECORD_SPEC = RecordSpecification.createSpecification("DummyTable", "DummyKey", 
            field("PostType", 2, false).numerical().doNotPersist(),
            field("OpgDato", 8, false),
            field("Timestamp", 20, false),
            field("Modt", 6, false),
            field("SnitfladeId", 8, false));
    
    public static final RecordSpecification END_RECORD_SPEC = RecordSpecification.createSpecification("DummyTable", "DummyKey", 
            field("PostType", 2, false).numerical().doNotPersist(),
            field("AntPost", 8, false).numerical());
    
    public static final RecordSpecification ENTRY_RECORD_SPEC = RecordSpecification.createSpecification("Sikrede", "CPRnr", 
            // Page 1 (6 of 11)
            field("PostType", 2, false).numerical().doNotPersist(),
            field("CPRnr", 10, false),
            field("SYdernr", 6, false),
            field("SIkraftDatoYder", 8, false),
            field("SRegDatoYder", 8, false),
            field("SSikrGrpKode", 1, false),
            field("SIkraftDatoGrp", 8, false),
            field("SRegDatoGrp", 8, false),
            field("SSikrKomKode", 3, false),
            field("SIkraftDatoKomKode", 8, false),
            field("SYdernrGl", 6, false),
            field("SIkraftDatoYderGl", 8, false),
            field("SRegDatoYderGl", 8, false),
            field("SSikrGrpKodeGl", 1, false),
            field("SIkraftDatoGrpGl", 8, false),
            field("SRegDatoGrpGl", 8, false),
            field("SYdernrFrem", 6, false),
            field("SIkraftDatoYderFrem", 8, false),
            field("SRegDatoYderFrem", 8, false),
            field("SSikrGrpKodeFrem", 1, false),

            // Page 2 (7 of 11)
            field("SIkraftDatoGrpFrem", 8, false),
            field("SRegDatoGrpFrem", 8, false),
            field("SKon", 1, false),
            field("SAlder", 3, false),
            field("SFolgerskabsPerson", 10, false),
            field("SStatus", 2, false),
            field("SBevisDato", 8, false),
            // ...
            field("PNavn", 34, false),
            // ...
            field("SBSStatsborgerskabKode", 2, false),
            field("SBSStatsborgerskab", 47, false),
            field("SSKAdrLinie1", 40, false),
            field("SSKAdrLinie2", 40, false),

            // Page 3 (8 of 11)
            field("SSKBopelsLand", 40, false),
            field("SSKBopelsLAndKode", 2, false),
            field("SSKEmailAdr", 50, false),
            field("SSKFamilieRelation", 10, false),
            field("SSKFodselsdato", 10, false),
            field("SSKGyldigFra", 10, false),
            field("SSKGyldigTil", 10, false),
            field("SSKMobilNr", 20, false),
            field("SSKPostNrBy", 40, false),
            field("SSLForsikringsinstans", 21, false),
            field("SSLForsikringsinstansKode", 10, false),
            field("SSLForsikringsnr", 15, false),
            field("SSLGyldigFra", 10, false),
            field("SSLGyldigTil", 10, false),
            field("SSLSocSikretLand", 47, false),
            field("SSLSocSikretLandKode", 2, false));
}
